# <center><u>**`Data Cleaning`**</u></center>

 Data cleaning is a part of the process on a data science project.
 
Data cleaning is the `process of fixing or removing incorrect, corrupted, incorrectly formatted, duplicate, or incomplete data `within a dataset.
<br>
When you clean your data, all outdated or incorrect information is gone – leaving you with the highest quality information for your analysis and model building.

<p style='text-align: right;'> 4 points</p>


In [162]:
# ignore this code for now
from IPython.display import YouTubeVideo

# YouTubeVideo('https://www.youtube.com/watch?v=XvNCSpIE9PE&list=PLsR_0x6BuM-F4Z72h5D1bjWEASL0_JlrF&index=4',width=600,height=300)

In [2]:
# import pandas and numpy with alias pd and np respectively
import pandas as pd
import numpy as np


Create a dataframe, d1 = pd.DataFrame( {‘Temperature’ : [1, np.nan, 3, 2, 3] ,’Humidity’ : [22, np.nan, 2 , np.nan, 20 ] })

In [3]:
df = pd.DataFrame( {'Temperature' : [1, np.nan, 3, 2, 3] ,'Humidity' : [22, np.nan, 2 , np.nan, 20 ] })

print the dataframe d1

In [4]:
df

Temperature  Humidity
0          1.0      22.0
1          NaN       NaN
2          3.0       2.0
3          2.0       NaN
4          3.0      20.0

In [163]:
# YouTubeVideo('https://www.youtube.com/watch?v=7Q_KoX3vAZ8&list=PLsR_0x6BuM-F4Z72h5D1bjWEASL0_JlrF&index=3',width=600,height=300)

Find whether the given dataframe contain any missing values? 

In [6]:
df.isnull()

Temperature  Humidity
0        False     False
1         True      True
2        False     False
3        False      True
4        False     False

How many missing values does each column have?

In [7]:
column_names = list(df)
column_names
zero_count = (df[column_names] == 0).sum()
NaN_count = df.isna().sum()
zero_count,NaN_count

(Temperature    0
 Humidity       0
 dtype: int64,
 Temperature    1
 Humidity       2
 dtype: int64)

In [164]:
# YouTubeVideo('https://www.youtube.com/watch?v=C0hO3FiYxlk&list=PLsR_0x6BuM-F4Z72h5D1bjWEASL0_JlrF&index=2',width=600,height=300)

## **`Dealing with missing values`**
<br>Now, we know we have missing values, the next thing that we need to work on, is how to deal with these missing values

In [165]:
# YouTubeVideo('https://www.youtube.com/watch?v=p-w5QbfWpHU&list=PLsR_0x6BuM-F4Z72h5D1bjWEASL0_JlrF&index=1',width=600,height=300)

## **`Method 1: Delete the rows which contain missing values.`**
 This method include dropping all the rows that have missing value in any column. 

 <p style='text-align: right;'> 20 points</p>

Use a suitable method to drop all the rows having missing values and save the change in d2 variable

In [10]:
df1 =  df.dropna(inplace=False)
df1.reset_index(drop=True)

Temperature  Humidity
0          1.0      22.0
1          3.0       2.0
2          3.0      20.0

Print df1

In [11]:
print(df1)

   Temperature  Humidity
0          1.0      22.0
2          3.0       2.0
4          3.0      20.0


Hey Remember : droping rows with nan is one of the method to deal with missing values. But you have to decide if you need to go for this method by checking percentage of nan present in the dataframe.

If a column is having more than 60% of nan values then its better to remove such variables altogether if business permits 

- ## **`Method 2: Replacing missing values`**
Sometimes rather than dropping NA values, you'd rather replace them with a valid value. This value might be a single number like zero, or it might be some sort of imputation or interpolation

<p style='text-align: right;'> 30 points</p>


Impute the missing values with constant number of your choice

In [12]:
# The below output has imputed missing  values with 100
df[column_names].replace(np.nan , 100)



Temperature  Humidity
0          1.0      22.0
1        100.0     100.0
2          3.0       2.0
3          2.0     100.0
4          3.0      20.0

Do you think its a  good way to treat Nan values? What if such constant values are not suitable for our further analysis?
Try to give your thoughts on this.

### It is better to give some valid values instead of nan values

Impute the missing values with mean

In [13]:
#  imputing mean
df[column_names].replace(np.nan, df[column_names].mean())


Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  14.666667
4         3.00  20.000000

Impute the missing values with median

In [14]:
#median imputing
df[column_names].fillna(df[column_names].median())

Temperature  Humidity
0          1.0      22.0
1          2.5      20.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Replacing with the mean, mode or median approximations are a statistical approach of handling the missing values.

Another Fun fact:
    
    This is an approximation which can add variance to the data set. But the loss of the data can be negated by this method which yields better results compared to removal of rows and columns.

Impute Nan with forward fill

In [15]:
#forward fill
df[column_names].fillna( df[column_names].ffill())

Temperature  Humidity
0          1.0      22.0
1          1.0      22.0
2          3.0       2.0
3          2.0       2.0
4          3.0      20.0

Impute Nan with backward fill

In [16]:
#backward fill
df.fillna( df[column_names].bfill())

Temperature  Humidity
0          1.0      22.0
1          3.0       2.0
2          3.0       2.0
3          2.0      20.0
4          3.0      20.0

Hey a fun fact here, as sweet as a cookie:

- ffill/pad/bfill are good imputation method if our data is of time series. This would keep the trend unaffected for our analysis.

Impute nan using interpolation method

In [17]:
#interpolate
(df.interpolate())

Temperature  Humidity
0          1.0      22.0
1          2.0      12.0
2          3.0       2.0
3          2.0      11.0
4          3.0      20.0

You lucky champ! you got to know another amazing fact:
 - Interpolation method by default is linear in nature. It is an imputation technique that assumes a linear relationship between data points and utilises non-missing values from adjacent data points to compute a value for a missing data point.


You can explore other techniques involved in interplolation method, which might be usefull for your project.

Perform KNN imputation


In [18]:
# Hint: Import KNNImputer and impute it on d1. Also note: Use n_neighbors=2
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=2)
new_df = pd.DataFrame(imputer.fit_transform(df),columns=('Temperature','Humidity'))
new_df

Temperature   Humidity
0         1.00  22.000000
1         2.25  14.666667
2         3.00   2.000000
3         2.00  12.000000
4         3.00  20.000000

Point to ponder: KNN is an algorithm that is useful for matching a point with its closest k neighbors in a multi-dimensional space.

Do you think scaling is required to implement this method?. Yes you are right the answer is YES.
Can you comment below why normalized data is required, so that we understand your logic on this.

It requires us to normalize our data. Otherwise, the different scales of our data will lead the KNN Imputer to generate biased replacements for the missing values.

## **`Dropping Irrelevant Columns`**


<p style='text-align: right;'> 5 points</p>


Create a dataframe df = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE'))


In [19]:
np.random.seed(10)

df_req = pd.DataFrame(np.random.randint(0,100,size=(100, 5)), columns=list('ABCDE'))
df_req.head()

A   B   C   D   E
0   9  15  64  28  89
1  93  29   8  73   0
2  40  36  16  11  54
3  88  62  33  72  78
4  49  51  54  77  69

print df

In [20]:
df_req

A   B   C   D   E
0    9  15  64  28  89
1   93  29   8  73   0
2   40  36  16  11  54
3   88  62  33  72  78
4   49  51  54  77  69
..  ..  ..  ..  ..  ..
95   3  50  59  34  21
96  16  18  61  54  60
97  21  87  83  71  16
98  67  38  27  96  87
99  98  89  16  82  19

[100 rows x 5 columns]

Note: Since all the rows are having random numbers, your dataframe observations might be different than the output given above

Suppose for our analysis our project do not require column E. So you need to remove this column. update this new change using inplace parameter

In [21]:
df_req.drop('E' , axis = 1, inplace=True)

Check if column **E** is removed by printing head of df

In [22]:
df_req.head()

A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

### `Ensure requirements as per domain`

<p style='text-align: right;'> 10 points</p>


In [166]:
# YouTubeVideo('https://www.youtube.com/watch?v=Y230Di2SSz8&list=PLsR_0x6BuM-F4Z72h5D1bjWEASL0_JlrF&index=5',width=600,height=300)


Shallow copy the dataframe df in variable df2 and print df2 head






In [24]:
import copy

df2_req = df_req.copy()
df2_req.head()
#df2 head


A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77

Suppose your domain expert says to filter column B with even numbers to do correct analysis. Implement the same below and update the change in varaible df2.

In [25]:
df2_req = df2_req[df2_req.B%2 == 0]

print updated head of *df2*

In [26]:
# df2 head
df2_req.head()

A   B   C   D
2   40  36  16  11
3   88  62  33  72
6   30  30  89  12
9   11  28  74  88
10  15  18  80  71

### `Creating a sensible index values`



Oops. The index in this dataframe doesn't make sense. please correct index in a sequential manner starting from 1. Save the updates in df2

In [27]:
#set proper index
df2_req.reset_index(inplace = True , drop=True)

print df2 head again

In [28]:
# df2 head
df2_req.head()

A   B   C   D
0  40  36  16  11
1  88  62  33  72
2  30  30  89  12
3  11  28  74  88
4  15  18  80  71

## **`Renaming column names to meaningful names.`**

<p style='text-align: right;'> 2 points</p>


Now df2 columns represents marks of the adventurous 'Anand', the brave 'Barkha', the compassionate 'Chandu' and the dashing 'Daniel'. Rename the columns with their name inplace of their first letter of their name.

In [29]:
#column renamingta
df2_req = df2_req.rename(columns={"A":"Anand" , "B":"Barkha", "C":"Chandu", "D":"Daniel"})
df2_req.tail()

Anand  Barkha  Chandu  Daniel
43      1      82      34      11
44     74      36       6      63
45      3      50      59      34
46     16      18      61      54
47     67      38      27      96

print df2 tail with updated column names

In [30]:
# df2 head
df2_req.head()

Anand  Barkha  Chandu  Daniel
0     40      36      16      11
1     88      62      33      72
2     30      30      89      12
3     11      28      74      88
4     15      18      80      71

Yeah! now the data looks pretty meaningful to study

## **`Treating Duplicate Data`**


<p style='text-align: right;'> 20 points</p>


Make another dataframe df3 by deep copying df2.

In [31]:
df3_req = df2_req.copy()

Make another column in df3 with name 'dummy' having 0 as values
throughout the rows.

In [32]:
#assign dummy column full of zero value
df3_req['dummy'] = df3_req['Anand'].apply(lambda x : x - x) 
df3_req.head()

Anand  Barkha  Chandu  Daniel  dummy
0     40      36      16      11      0
1     88      62      33      72      0
2     30      30      89      12      0
3     11      28      74      88      0
4     15      18      80      71      0

In [33]:
# print head of df, df2 and df3
print(df_req.head())
print(df2_req.head())
print(df3_req.head())
print(df_req.tail())
print(df2_req.tail())
print(df3_req.tail())

    A   B   C   D
0   9  15  64  28
1  93  29   8  73
2  40  36  16  11
3  88  62  33  72
4  49  51  54  77
   Anand  Barkha  Chandu  Daniel
0     40      36      16      11
1     88      62      33      72
2     30      30      89      12
3     11      28      74      88
4     15      18      80      71
   Anand  Barkha  Chandu  Daniel  dummy
0     40      36      16      11      0
1     88      62      33      72      0
2     30      30      89      12      0
3     11      28      74      88      0
4     15      18      80      71      0
     A   B   C   D
95   3  50  59  34
96  16  18  61  54
97  21  87  83  71
98  67  38  27  96
99  98  89  16  82
    Anand  Barkha  Chandu  Daniel
43      1      82      34      11
44     74      36       6      63
45      3      50      59      34
46     16      18      61      54
47     67      38      27      96
    Anand  Barkha  Chandu  Daniel  dummy
43      1      82      34      11      0
44     74      36       6      63      0
45      3    

Hey buddy! Don't you think, there is some difference between copy operation used for creating df2 and df3.

If you think Yes, Then please comment below the difference

In [34]:
# comment

print tail of df3

In [35]:
df3_req.tail()

Anand  Barkha  Chandu  Daniel  dummy
43      1      82      34      11      0
44     74      36       6      63      0
45      3      50      59      34      0
46     16      18      61      54      0
47     67      38      27      96      0

make an array name 'ListB' with values of column 'Barkha'

In [36]:
ListB=list(df3_req.Barkha)

Print ListB

In [37]:
#print ListB
print(ListB)

[36, 62, 30, 28, 18, 50, 88, 50, 80, 66, 96, 30, 4, 30, 2, 42, 94, 18, 44, 68, 58, 48, 70, 22, 36, 32, 32, 96, 30, 86, 0, 76, 88, 64, 52, 46, 20, 66, 56, 8, 68, 50, 28, 82, 36, 50, 18, 38]


Assign this array values as another column in df3 with name 'Anonymous'

In [38]:
#create Anonymous column
df3_req['Anonymous'] = ListB
df3_req.head()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
0     40      36      16      11      0         36
1     88      62      33      72      0         62
2     30      30      89      12      0         30
3     11      28      74      88      0         28
4     15      18      80      71      0         18

Create a dataframe 'ListA' with values of row index 3, 10 and 40

In [39]:
ListA=pd.DataFrame(df3_req.iloc[[3,10,40]])

print ListA

In [40]:
# print ListA
ListA

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
3      11      28      74      88      0         28
10     88      96      73      40      0         96
40     87      68      86      14      0         68

Concat ListA to df3 ignoring the index values of ListA so that we can maintain the sequential index value thoughout the dataframe.

In [41]:
df3_req = pd.concat([df3_req, ListA], axis=0).reset_index(inplace=False, drop = True)

print head of df3

In [42]:
# df3 head
df3_req.tail()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
46     16      18      61      54      0         18
47     67      38      27      96      0         38
48     11      28      74      88      0         28
49     88      96      73      40      0         96
50     87      68      86      14      0         68

Check if there is any duplicate rows present in the dataframe df3

In [43]:
#check duplicate
df3_req[df3_req.duplicated()]

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
48     11      28      74      88      0         28
49     88      96      73      40      0         96
50     87      68      86      14      0         68

By above output it seems we do have duplicated rows in our dataset

Drop duplicated rows using pandas function keeping first values of such duplicated observations

In [44]:
#drop duplicate
df3_req.drop_duplicates(inplace=True)
df3_req.tail()

Anand  Barkha  Chandu  Daniel  dummy  Anonymous
43      1      82      34      11      0         82
44     74      36       6      63      0         36
45      3      50      59      34      0         50
46     16      18      61      54      0         18
47     67      38      27      96      0         38

Check again if we have any duplicate row values present 

In [45]:
#check duplicate
df3_req[df3_req.duplicated()]

Empty DataFrame
Columns: [Anand, Barkha, Chandu, Daniel, dummy, Anonymous]
Index: []

Yipeee!! Did you notice the dataframe is free from any duplicate rows now.

Drop any duplicate columns present in the dataframe df

In [46]:
df3_req = df3_req.T.drop_duplicates().T

#print df3
df3_req

Anand  Barkha  Chandu  Daniel  dummy
0      40      36      16      11      0
1      88      62      33      72      0
2      30      30      89      12      0
3      11      28      74      88      0
4      15      18      80      71      0
5      88      50      54      34      0
6      77      88      15       6      0
7      97      50      45      40      0
8      81      80      41      90      0
9      96      66      67      62      0
10     88      96      73      40      0
11     28      30      89      25      0
12     33       4      87      94      0
13     68      30      70      74      0
14      9       2      65      13      0
15     62      42      34      40      0
16     32      94      86      58      0
17     45      18      50      44      0
18      6      44       9      50      0
19     44      68      14       4      0
20     39      58      81      39      0
21     69      48      60      58      0
22      5      70      61      16      0
23      4      22      19      85      0
24     10      36       0      30      0
25     85      32      81      13      0
26     31      32      96      38      0
27      0      96      25      63      0
28      2      30      17      59      0
29     63      86      29      71      0
30     19       0       7      93      0
31     58      76      10      79      0
32      7      88      96      38      0
33     27      64      49      90      0
34     27      52      90      57      0
35     99      46      46      45      0
36     84      20      49      33      0
37     77      66      37      31      0
38     82      56      41      90      0
39     74       8      92      32      0
40     87      68      86      14      0
41     50      50      85      14      0
42      1      28      58      10      0
43      1      82      34      11      0
44     74      36       6      63      0
45      3      50      59      34      0
46     16      18      61      54      0
47     67      38      27      96      0

Did you notice which Column is dropped?
I am sure you noticed it.
Name that column below

In [47]:
#Column: Anonymous

### `Treating constant column values`

<p style='text-align: right;'> 2 points</p>


Check unique values in each columns

In [48]:
# df3 unique value
print(df3_req.Anand.unique())
print(df3_req.Barkha.unique())
print(df3_req.Chandu.unique())
print(df3_req.Daniel.unique())
print(df3_req.dummy.unique())

[40 88 30 11 15 77 97 81 96 28 33 68  9 62 32 45  6 44 39 69  5  4 10 85
 31  0  2 63 19 58  7 27 99 84 82 74 87 50  1  3 16 67]
[36 62 30 28 18 50 88 80 66 96  4  2 42 94 44 68 58 48 70 22 32 86  0 76
 64 52 46 20 56  8 82 38]
[16 33 89 74 80 54 15 45 41 67 73 87 70 65 34 86 50  9 14 81 60 61 19  0
 96 25 17 29  7 10 49 90 46 37 92 85 58  6 59 27]
[11 72 12 88 71 34  6 40 90 62 25 94 74 13 58 44 50  4 39 16 85 30 38 63
 59 93 79 57 45 33 31 32 14 10 54 96]
[0]


By above output which column has only 1 value as unique throught the rows? 
Yeah! you are right, its dummy column.
So lets drop it

Drop dummy column as it has constant values which will not give us any information and save the changes to df3 using inplace parameter

In [49]:
# drop dummy
df3_req.drop('dummy',axis=1,inplace=True)

print final obtained dataframe df3

In [50]:
# print df3
df3_req

Anand  Barkha  Chandu  Daniel
0      40      36      16      11
1      88      62      33      72
2      30      30      89      12
3      11      28      74      88
4      15      18      80      71
5      88      50      54      34
6      77      88      15       6
7      97      50      45      40
8      81      80      41      90
9      96      66      67      62
10     88      96      73      40
11     28      30      89      25
12     33       4      87      94
13     68      30      70      74
14      9       2      65      13
15     62      42      34      40
16     32      94      86      58
17     45      18      50      44
18      6      44       9      50
19     44      68      14       4
20     39      58      81      39
21     69      48      60      58
22      5      70      61      16
23      4      22      19      85
24     10      36       0      30
25     85      32      81      13
26     31      32      96      38
27      0      96      25      63
28      2      30      17      59
29     63      86      29      71
30     19       0       7      93
31     58      76      10      79
32      7      88      96      38
33     27      64      49      90
34     27      52      90      57
35     99      46      46      45
36     84      20      49      33
37     77      66      37      31
38     82      56      41      90
39     74       8      92      32
40     87      68      86      14
41     50      50      85      14
42      1      28      58      10
43      1      82      34      11
44     74      36       6      63
45      3      50      59      34
46     16      18      61      54
47     67      38      27      96

### `Iterating dataframes`
<p style='text-align: right;'> 25 points </p>


Let's look at three main ways to iterate over DataFrames.

1. iteritems()
2. iterrows()
3. itertuples()

We will also see time taken by these methods to print our dataframe. 

**1. Iterating DataFrames with iteritems()**

Lets iterate over rows of df3 uisng iteritems.



In [167]:
# YouTubeVideo('https://www.youtube.com/watch?v=rl3Nk-IxXV4&list=PLsR_0x6BuM-F4Z72h5D1bjWEASL0_JlrF&index=6',width=600,height=300)

In [168]:
import time
start = time.time()

#Use iteritems to iterate

for index in range(5):
    print(index, df3_req['Anand'] [index], df3_req['Chandu'][index])

# print('Time taken(sec): ',(time.time()-start)*1000)
# start

0 40 16
1 88 33
2 30 89
3 11 74
4 15 80


In [53]:
# access using loc
for index in range(5):
    print(df3_req.loc[index,'Anand'], df3_req.loc[index,'Chandu'])


40 16
88 33
30 89
11 74
15 80


In [54]:
# access using iloc
for index in range(5):
    print(df3_req.iloc[index,0], df3_req.iloc[index,1])

40 36
88 62
30 30
11 28
15 18


In [55]:
# Access using apply()
df3_req.apply(lambda x: str(x['Anand']) + " " +  str(x['Chandu']) , axis=1).head()

0    40 16
1    88 33
2    30 89
3    11 74
4    15 80
dtype: object

In [56]:
# iteritems -> printing columnwise
for column,data in df3_req.head().iteritems():
    print(column , '\n' , data ,'\n', '*****' *10)

Anand 
 0    40
1    88
2    30
3    11
4    15
Name: Anand, dtype: int64 
 **************************************************
Barkha 
 0    36
1    62
2    30
3    28
4    18
Name: Barkha, dtype: int64 
 **************************************************
Chandu 
 0    16
1    33
2    89
3    74
4    80
Name: Chandu, dtype: int64 
 **************************************************
Daniel 
 0    11
1    72
2    12
3    88
4    71
Name: Daniel, dtype: int64 
 **************************************************


Did you notice buddy how iteritems are iterating over df3.

Along with ways each iterating function works, also keep tallying the time taken for all other lopps too!. This will be fun, lets check iterrows()

**2. Iterating DataFrames with iterrows()**

In [57]:
import time
start = time.time()
#Use iterrows to iterate
# iterrows -> printing rowwise
for index,data in df3_req.head().iterrows():
    print(index , '\n' , data ,'\n', '*****' *10)
print('Time taken(sec): ',(time.time()-start)*1000)

0 
 Anand     40
Barkha    36
Chandu    16
Daniel    11
Name: 0, dtype: int64 
 **************************************************
1 
 Anand     88
Barkha    62
Chandu    33
Daniel    72
Name: 1, dtype: int64 
 **************************************************
2 
 Anand     30
Barkha    30
Chandu    89
Daniel    12
Name: 2, dtype: int64 
 **************************************************
3 
 Anand     11
Barkha    28
Chandu    74
Daniel    88
Name: 3, dtype: int64 
 **************************************************
4 
 Anand     15
Barkha    18
Chandu    80
Daniel    71
Name: 4, dtype: int64 
 **************************************************
Time taken(sec):  3.0045509338378906


**3. Iterating DataFrames with itertuples()**

In [58]:
#iterate df3 using itertuples
import time
start = time.time()

#Use itertuples to iterate

# iterrows -> printing like  tuples
for index in df3_req.head().itertuples():
    print(index , '\n')


print('Time taken(sec): ',(time.time()-start)*1000)

Pandas(Index=0, Anand=40, Barkha=36, Chandu=16, Daniel=11) 

Pandas(Index=1, Anand=88, Barkha=62, Chandu=33, Daniel=72) 

Pandas(Index=2, Anand=30, Barkha=30, Chandu=89, Daniel=12) 

Pandas(Index=3, Anand=11, Barkha=28, Chandu=74, Daniel=88) 

Pandas(Index=4, Anand=15, Barkha=18, Chandu=80, Daniel=71) 

Time taken(sec):  1.0013580322265625


In [59]:
#  iterate by column
for column_name in df3_req.columns:
    print(column_name, '\n')
    print(df3_req[column_name].head())

Anand 

0    40
1    88
2    30
3    11
4    15
Name: Anand, dtype: int64
Barkha 

0    36
1    62
2    30
3    28
4    18
Name: Barkha, dtype: int64
Chandu 

0    16
1    33
2    89
3    74
4    80
Name: Chandu, dtype: int64
Daniel 

0    11
1    72
2    12
3    88
4    71
Name: Daniel, dtype: int64


Hey buddy! so as you have seen every method works differently 
    
    iteritems(): Helps to iterate over each element of the set, column-wise. 
    iterrows(): Each element of the set, row-wise. 
    itertuple(): Each row and form a tuple out of them.

But if you ask for speed. The most best perfromance is given by itertuples compared to other two iterating methods.
So if anytime you need to save your computation time on iteration of dataframes you can go for itertuples. Was'nt it fun?:)

### `Regular Expression` 
<p style='text-align: right;'> 15 points </p>


Reference: Watch the video below

Reference doc: https://www.guru99.com/python-regular-expressions-complete-tutorial.html

Python has a module named re to work with RegEx


### **!Are you ready to try regex on dataframes?**

*So here we go.!*

We are gonna try out following awesome re module functions

1. findall 
2. search
3. sub
4. split

If you want you can also refer the below regular expression syntax.

![image.png](attachment:image.png)


In [169]:
# YouTubeVideo('YXFZTKlkf-Y',width=600,height=300)

Hey future data scientists! we will now use regex on dataframes for data cleaning.

Who doesn't know Trump?. Lets dowload this interesting dataset of Trump insult tweets :https://www.kaggle.com/ayushggarg/all-trumps-twitter-insults-20152021/download

On this dataset we will learn how to use regex for data cleaning. By the way it will be also very usefull for feature engineering too!.



In [269]:
#load dataset
tweet_data=pd.read_csv('trump_insult_tweets_2014_to_2021.csv')
tweet_data.reset_index(inplace=True , drop=True)
tweet_data.drop('Unnamed: 0' , axis=1,inplace=True)
tweet_data.head()

date          target  \
0  2014-10-09  thomas-frieden   
1  2014-10-09  thomas-frieden   
2  2015-06-16     politicians   
3  2015-06-24      ben-cardin   
4  2015-06-24      neil-young   

                                              insult  \
0                                               fool   
1                                               DOPE   
2                             all talk and no action   
3  It's politicians like Cardin that have destroy...   
4                                    total hypocrite   

                                               tweet  
0  Can you believe this fool, Dr. Thomas Frieden ...  
1  Can you believe this fool, Dr. Thomas Frieden ...  
2  Big time in U.S. today - MAKE AMERICA GREAT AG...  
3  Politician @SenatorCardin didn't like that I s...  
4  For the nonbeliever, here is a photo of @Neily...

In [270]:
#import re module
import re

#print head of tweet_data
tweet_data.head()

date          target  \
0  2014-10-09  thomas-frieden   
1  2014-10-09  thomas-frieden   
2  2015-06-16     politicians   
3  2015-06-24      ben-cardin   
4  2015-06-24      neil-young   

                                              insult  \
0                                               fool   
1                                               DOPE   
2                             all talk and no action   
3  It's politicians like Cardin that have destroy...   
4                                    total hypocrite   

                                               tweet  
0  Can you believe this fool, Dr. Thomas Frieden ...  
1  Can you believe this fool, Dr. Thomas Frieden ...  
2  Big time in U.S. today - MAKE AMERICA GREAT AG...  
3  Politician @SenatorCardin didn't like that I s...  
4  For the nonbeliever, here is a photo of @Neily...

Lets do some analysis using regex on this dataset

Before we go ahead, do you remember apply function? because you will have to require apply function to impliment regex methods.

You can refer video below:
**1. findall()**

Make another column 'year' with year in each row using regex on date column.


In [271]:
# YouTubeVideo('T-b7TxWfrfE',width=600,height=300)

In [272]:
# create a function which takes date as parameter and applies regex on it


#use apply function on tweet_data to use above function in order to make year column


#print tweet_data head


def split_it(year):
    pattern = r'^20\d\d'
    return str(re.findall(pattern, year)[0])

tweet_data['Year'] = tweet_data['date'].apply(lambda x: split_it(x))
tweet_data.head()

date          target  \
0  2014-10-09  thomas-frieden   
1  2014-10-09  thomas-frieden   
2  2015-06-16     politicians   
3  2015-06-24      ben-cardin   
4  2015-06-24      neil-young   

                                              insult  \
0                                               fool   
1                                               DOPE   
2                             all talk and no action   
3  It's politicians like Cardin that have destroy...   
4                                    total hypocrite   

                                               tweet  Year  
0  Can you believe this fool, Dr. Thomas Frieden ...  2014  
1  Can you believe this fool, Dr. Thomas Frieden ...  2014  
2  Big time in U.S. today - MAKE AMERICA GREAT AG...  2015  
3  Politician @SenatorCardin didn't like that I s...  2015  
4  For the nonbeliever, here is a photo of @Neily...  2015

lets filter year from 2020-2021 which was the election time in USA.

**2. search**

We will use regex search for this 

In [273]:
# create a function which takes year as parameter and applies regex on it
def filter_year(year):
    x= re.search( r'^202\d' , year)
    if x :
        return x.group(0)
    
#use apply function on tweet_data to use above function in order to search tweets of 2020-2021
tweet_data_search = tweet_data
tweet_data_search['Year'] = tweet_data_search['Year'].apply(filter_year)


# tweet_data_new = tweet_data[(tweet_data.Year == 2020) | (tweet_data.Year == 2021)]
# tweet_data_new.reset_index(inplace=True, drop=True)
# tweet_data_new.head()
# #tweet_data head


In [274]:
tweet_data_search.tail(5)

date         target  \
10355  2021-01-06  2020-election   
10356  2021-01-06  2020-election   
10357  2021-01-06  2020-election   
10358  2021-01-06  2020-election   
10359  2021-01-06     chuck-todd   

                                                  insult  \
10355  Many States want to decertify the mistake they...   
10356  based on irregularities and fraud, plus corrup...   
10357  Our Election Process is worse than that of thi...   
10358                                            a FRAUD   
10359                        Sleepy Eyes,  Sad to watch!   

                                                   tweet  Year  
10355  If Vice President @Mike_Pence comes through fo...  2021  
10356  States want to correct their votes, which they...  2021  
10357  They just happened to find 50,000 ballots late...  2021  
10358  The States want to redo their votes. They foun...  2021  
10359  Sleepy Eyes Chuck Todd is so happy with the fa...  2021

You can also do the same thing using regex match function to do this which is vailable in pandas

Reference: https://www.geeksforgeeks.org/python-pandas-series-str-match/

In [275]:
# apply pandas str.match() function
tweet_data_match = tweet_data
tweet_data_match.Year = tweet_data_match.Year.str.match(pat = r'^202\d')



In [276]:
tweet_data_match.head(5)

date          target  \
0  2014-10-09  thomas-frieden   
1  2014-10-09  thomas-frieden   
2  2015-06-16     politicians   
3  2015-06-24      ben-cardin   
4  2015-06-24      neil-young   

                                              insult  \
0                                               fool   
1                                               DOPE   
2                             all talk and no action   
3  It's politicians like Cardin that have destroy...   
4                                    total hypocrite   

                                               tweet  Year  
0  Can you believe this fool, Dr. Thomas Frieden ...  None  
1  Can you believe this fool, Dr. Thomas Frieden ...  None  
2  Big time in U.S. today - MAKE AMERICA GREAT AG...  None  
3  Politician @SenatorCardin didn't like that I s...  None  
4  For the nonbeliever, here is a photo of @Neily...  None

cool right!

You got some null values after applying above function. Lets drop them using dropna function. Also drop 'Unnamed: 0' column as it does not give any information.


In [277]:
#drop na and Unnamed: 0 column

tweet_data_match.dropna(inplace=True)
tweet_data_match.reset_index(inplace = True , drop=True)
tweet_data_match

date               target  \
0     2020-01-01  impeachment-inquiry   
1     2020-01-02  impeachment-inquiry   
2     2020-01-04  impeachment-inquiry   
3     2020-01-06          los-angeles   
4     2020-01-06  impeachment-inquiry   
...          ...                  ...   
2734  2021-01-06        2020-election   
2735  2021-01-06        2020-election   
2736  2021-01-06        2020-election   
2737  2021-01-06        2020-election   
2738  2021-01-06           chuck-todd   

                                                 insult  \
0               The greatest Witch Hunt in U.S. history   
1     Witch Hunt is sputtering badly, but still goin...   
2                 this ridiculous Impeachment Lite Hoax   
3                                            poorly run   
4                              The great Scam continues   
...                                                 ...   
2734  Many States want to decertify the mistake they...   
2735  based on irregularities and fraud, plus corrup...   
2736  Our Election Process is worse than that of thi...   
2737                                            a FRAUD   
2738                        Sleepy Eyes,  Sad to watch!   

                                                  tweet  Year  
0     Thank you Steve. The greatest Witch Hunt in U....  True  
1     A lot of very good people were taken down by a...  True  
2     As hard as I work, & as successful as our Coun...  True  
3     ....If however, the city or state in question ...  True  
4     “The reason they are not sending the Articles ...  True  
...                                                 ...   ...  
2734  If Vice President @Mike_Pence comes through fo...  True  
2735  States want to correct their votes, which they...  True  
2736  They just happened to find 50,000 ballots late...  True  
2737  The States want to redo their votes. They foun...  True  
2738  Sleepy Eyes Chuck Todd is so happy with the fa...  True  

[2739 rows x 5 columns]

**3. sub() Function**

Now you have filtered the dataset with 2739 rows. Let's remove all @ from tweet column suing re sub() function.

In [278]:
# create a function which takes tweet as parameter and applies regex on it
def use_sub(tweet):
    return re.sub(r'@' , '',tweet)


#use apply function on tweet_data to use above function in order to remove @ from tweets
tweet_data_match['tweet'] = tweet_data_match['tweet'].apply(use_sub)


#tweet_data head
tweet_data_match

date               target  \
0     2020-01-01  impeachment-inquiry   
1     2020-01-02  impeachment-inquiry   
2     2020-01-04  impeachment-inquiry   
3     2020-01-06          los-angeles   
4     2020-01-06  impeachment-inquiry   
...          ...                  ...   
2734  2021-01-06        2020-election   
2735  2021-01-06        2020-election   
2736  2021-01-06        2020-election   
2737  2021-01-06        2020-election   
2738  2021-01-06           chuck-todd   

                                                 insult  \
0               The greatest Witch Hunt in U.S. history   
1     Witch Hunt is sputtering badly, but still goin...   
2                 this ridiculous Impeachment Lite Hoax   
3                                            poorly run   
4                              The great Scam continues   
...                                                 ...   
2734  Many States want to decertify the mistake they...   
2735  based on irregularities and fraud, plus corrup...   
2736  Our Election Process is worse than that of thi...   
2737                                            a FRAUD   
2738                        Sleepy Eyes,  Sad to watch!   

                                                  tweet  Year  
0     Thank you Steve. The greatest Witch Hunt in U....  True  
1     A lot of very good people were taken down by a...  True  
2     As hard as I work, & as successful as our Coun...  True  
3     ....If however, the city or state in question ...  True  
4     “The reason they are not sending the Articles ...  True  
...                                                 ...   ...  
2734  If Vice President Mike_Pence comes through for...  True  
2735  States want to correct their votes, which they...  True  
2736  They just happened to find 50,000 ballots late...  True  
2737  The States want to redo their votes. They foun...  True  
2738  Sleepy Eyes Chuck Todd is so happy with the fa...  True  

[2739 rows x 5 columns]

You can also use the sub function just in one line using list comprehension. Can you try doing it below?

In [279]:
# sub() suing list comprehension
tweet_data_match['tweet'] = tweet_data_match['tweet'].apply(lambda x : re.sub(r'&' , 'and',x))
tweet_data_match

date               target  \
0     2020-01-01  impeachment-inquiry   
1     2020-01-02  impeachment-inquiry   
2     2020-01-04  impeachment-inquiry   
3     2020-01-06          los-angeles   
4     2020-01-06  impeachment-inquiry   
...          ...                  ...   
2734  2021-01-06        2020-election   
2735  2021-01-06        2020-election   
2736  2021-01-06        2020-election   
2737  2021-01-06        2020-election   
2738  2021-01-06           chuck-todd   

                                                 insult  \
0               The greatest Witch Hunt in U.S. history   
1     Witch Hunt is sputtering badly, but still goin...   
2                 this ridiculous Impeachment Lite Hoax   
3                                            poorly run   
4                              The great Scam continues   
...                                                 ...   
2734  Many States want to decertify the mistake they...   
2735  based on irregularities and fraud, plus corrup...   
2736  Our Election Process is worse than that of thi...   
2737                                            a FRAUD   
2738                        Sleepy Eyes,  Sad to watch!   

                                                  tweet  Year  
0     Thank you Steve. The greatest Witch Hunt in U....  True  
1     A lot of very good people were taken down by a...  True  
2     As hard as I work, and as successful as our Co...  True  
3     ....If however, the city or state in question ...  True  
4     “The reason they are not sending the Articles ...  True  
...                                                 ...   ...  
2734  If Vice President Mike_Pence comes through for...  True  
2735  States want to correct their votes, which they...  True  
2736  They just happened to find 50,000 ballots late...  True  
2737  The States want to redo their votes. They foun...  True  
2738  Sleepy Eyes Chuck Todd is so happy with the fa...  True  

[2739 rows x 5 columns]

**4. split() Function**

Lets now split the column target by making another column Name which has name before hyphen("-")


In [280]:
# create a function which takes target as parameter and applies regex on it
def use_split(target):
    return re.split(r'-',target)[0]

#use apply function on tweet_data to use above function in order to create Name column

tweet_data_match['Name'] = tweet_data_match['target'].apply(use_split)
#tweet_data head
tweet_data_match.head()

date               target  \
0  2020-01-01  impeachment-inquiry   
1  2020-01-02  impeachment-inquiry   
2  2020-01-04  impeachment-inquiry   
3  2020-01-06          los-angeles   
4  2020-01-06  impeachment-inquiry   

                                              insult  \
0            The greatest Witch Hunt in U.S. history   
1  Witch Hunt is sputtering badly, but still goin...   
2              this ridiculous Impeachment Lite Hoax   
3                                         poorly run   
4                           The great Scam continues   

                                               tweet  Year         Name  
0  Thank you Steve. The greatest Witch Hunt in U....  True  impeachment  
1  A lot of very good people were taken down by a...  True  impeachment  
2  As hard as I work, and as successful as our Co...  True  impeachment  
3  ....If however, the city or state in question ...  True          los  
4  “The reason they are not sending the Articles ...  True  impeachment

Now can you filter out Name which are specially targetted for trump? Lets do it below and check how many such tweets are there.

In [281]:
#filter Name which is euqals to trump
def filter_trump(name):
    name = re.search(r'trump' , name)
    if name:
        name.group(0)

tweet_data_match = tweet_data_match[tweet_data_match.Name == 'trump']

In [285]:

tweet_data_match.reset_index(inplace=True , drop=True)
tweet_data_match

date         target  \
0   2020-02-09   trump-russia   
1   2020-02-09   trump-russia   
2   2020-02-09   trump-russia   
3   2020-02-12   trump-russia   
4   2020-02-12   trump-russia   
..         ...            ...   
60  2020-10-07   trump-russia   
61  2020-10-07   trump-russia   
62  2020-11-22   trump-russia   
63  2020-12-09  trump-critics   
64  2020-12-18   trump-russia   

                                               insult  \
0   THE WHOLE SCAM INVESTIGATION, THE MUELLER REPO...   
1   the biggest political crime in American Histor...   
2   THE PARTY IN POWER ILLEGALLY SPIED ON MY CAMPAIGN   
3                   an investigation that was illegal   
4                                    the Mueller Scam   
..                                                ...   
60                                       Hoax Scandal   
61  the biggest political crime in the history of ...   
62                            never ending Witch Hunt   
63                                          obnoxious   
64                                        Russia Hoax   

                                                tweet  Year   Name  
0   FBI Director Christopher Wray just admitted th...  True  trump  
1   ....This is the biggest political crime in Ame...  True  trump  
2   ....This is the biggest political crime in Ame...  True  trump  
3   Who are the four prosecutors (Mueller people?)...  True  trump  
4   Who are the four prosecutors (Mueller people?)...  True  trump  
..                                                ...   ...    ...  
60  All Russia Hoax Scandal information was Declas...  True  trump  
61  All Russia Hoax Scandal information was Declas...  True  trump  
62  Thanks Mark. It’s all a continuation of the ne...  True  trump  
63  Germany has consistently been used by my obnox...  True  trump  
64  The Russia Hoax becomes an even bigger lie! ht...  True  trump  

[65 rows x 6 columns]

So here you got total of 65 records which were  tweeted on Donald Trump in the span of 2020-2021. 

Well done buddy! You have learned how to apply regex on dataframes. Regex are mostly used for datasets which are having textual information.

Good job!  Now our interesting trump insult tweet data is somewhat cleaned. 

---------------------------------

# C'mon cheers:) you have completed the 5th milestone challenge too. 

--------------------------------

# FeedBack
We hope you’ve enjoyed this course so far. We’re committed to help you use "AI for All" course to its full potential, so that you have a great learning experience. And that’s why we need your help in form of a feedback here.

Please fill this feedback form https://docs.google.com/forms/d/e/1FAIpQLSfjBmH0yJSSA34IhSVx4h2eDMgOAeG4Dk-yHid__NMTk3Hq5g/viewform